utils: Add PCI device, drivers, modules, dmesg and vfio related avocado utilities

dhsrivas opened the pull request:
Add utilities to
1. Check whether the PCI device supports MSI-X and if it is currently enabled
2. Check if the device supports at least the specified number of interrupts (MSI-X vectors).
3. Add vendor id of a PCI device address "device" to "driver"
4. Unbind the given device from its current driver and bind it to the specified new driver.
5. Configure module
6. Check kernel logs

And utilities to use vfio ioctls.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Kernel log checking with fallback to the system journal.
  * Kernel module configuration with explicit outcomes and safe loading behavior.
  * PCI utilities for vendor registration, driver reassignment, MSI‑X capability detection, and IRQ-count checks.
  * VFIO utilities to manage containers, IOMMU groups, devices, and IRQ support queries.

* **Tests**
  * New unit tests for kernel logs, module configuration, PCI and VFIO utilities; updated unit-suite test count.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

clebergnu commented:
> @clebergnu @richtja Gentle reminder.

Hi @dhsrivas, I looked through the code, and they look like a great addition.  Thanks a lot for this contribution.

Having said that, you could improve this series by:

* adding some sort of testing.  I understand testing code that interacts with utilities and systems may require some mocking (in case of unittests) or elevated privileges (in case of functional tests) and those pose challenges when it comes to maintainability.  Still, it'd worth adding some, especially in view of the AAutils project (if you don't know about it, I can certainly give you pointers)
* (optional) splitting the commit with per subsystem/interest

Thanks again!

dhsrivas commented:
@clebergnu  Can you please point me some examples and where to add unit tests for these utils?

Also splitting of test case is fine to me. 

Thanks
Dheeraj

dhsrivas commented:
@clebergnu  Gentle reminder.

Thanks
Dheeraj

richtja reviewed (changes requested):
Hi @dhsrivas, thank you very much about introducing many new utilities, this effort is very welcome. I have a couple of comments to your changes, please let me know what do you think.

In addition to my comments, I have two general requests:
- Please resolve the current conflicts, for that is needed in every commit where you are adding tests you need to update the overall number in [check.py](https://github.com/avocado-framework/avocado/blob/master/selftests/check.py#L30)
- Remove the Merge commit 39dccbbbb5d3b06739933bc07ce1361114f9714f, it is not needed for this PR

dhsrivas commented on avocado/utils/pci.py:
```diff
@@ -376,6 +377,88 @@ def get_vendor_id(full_pci_address):
+    :param driver: Driver to be attached to the specified PCI device
+    """
+
+    try:
```
For a cleaner design, let the caller handle saving the initial PCI device state and restoring it on failure, rather than doing it in this utility.

dhsrivas commented:
Closing this PR and will send separate PRs. One for existing file utilities (pci.py, linux_modules.py, dmesg.py) and one for new utility file (vfio.py) addressing all review comments here. Thanks
